File: scientific_library/tvb/datatypes/annotations.py
# -*- coding: utf-8 -*-
#
#
# TheVirtualBrain-Framework Package. This package holds all Data Management, and
# Web-UI helpful to run brain-simulations. To use it, you also need do download
# TheVirtualBrain-Scientific Package (for simulators). See content of the
# documentation-folder for more details. See also http://www.thevirtualbrain.org
#
# (c) 2012-2017, Baycrest Centre for Geriatric Care ("Baycrest") and others
#
# This program is free software: you can redistribute it and/or modify it under the
# terms of the GNU General Public License as published by the Free Software Foundation,
# either version 3 of the License, or (at your option) any later version.
# This program is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
# PARTICULAR PURPOSE.  See the GNU General Public License for more details.
# You should have received a copy of the GNU General Public License along with this
# program.  If not, see <http://www.gnu.org/licenses/>.
#
#
#   CITATION:
# When using The Virtual Brain for scientific publications, please cite it as follows:
#
#   Paula Sanz Leon, Stuart A. Knock, M. Marmaduke Woodman, Lia Domide,
#   Jochen Mersmann, Anthony R. McIntosh, Viktor Jirsa (2013)
#       The Virtual Brain: a simulator of primate brain network dynamics.
#   Frontiers in Neuroinformatics (7:10. doi: 10.3389/fninf.2013.00010)
#
#

"""
.. moduleauthor:: Lia Domide <lia.domide@codemart.ro>
"""

import numpy
from tvb.datatypes.connectivity import Connectivity
from tvb.basic.neotraits.api import Attr, NArray, HasTraits, narray_summary_info

ANNOTATION_DTYPE = numpy.dtype([('id', 'i'),
                                ('parent_id', 'i'),
                                ('parent_left', 'i'),
                                ('parent_right', 'i'),
                                ('relation', 'S16'),
                                ('label', 'S128'),
                                ('definition', 'S1024'),
                                ('synonym', 'S2048'),
                                ('uri', 'S248'),
                                ('synonym_tvb_left', 'i'),
                                ('synonym_tvb_right', 'i')
                                ])

ICON_TVB = "/static/style/nodes/nodeRoot.png"
ICON_FOLDER = "/static/style/nodes/nodeFolder.png"
NODE_ID_TVB = "node_tvb_"
NODE_ID_TVB_ROOT = "node_tvb_root_"
NODE_ID_BRCO = "node_brco_"


class AnnotationTerm(object):
    """
    One single annotation node (in the tree of annotations / region)
    """

    def __init__(self, id, parent, parent_left, parent_right, relation, label, definition=None, synonym=None, uri=None,
                 tvb_left=None, tvb_right=None):
        self.id = id
        self.parent_id = parent
        self.parent_left = parent_left
        self.parent_right = parent_right
        self.relation = relation
        self.label = label
        self.definition = definition
        self.synonym = synonym
        self.uri = uri
        self.synonym_tvb_left = tvb_left or -1
        self.synonym_tvb_right = tvb_right or -1
        self.children = []

    def add_child(self, annotation_child):
        self.children.append(annotation_child)

    def to_tuple(self):
        return self.id, self.parent_id, self.parent_left, self.parent_right, \
               self.relation, self.label, self.definition, self.synonym, self.uri, \
               self.synonym_tvb_left, self.synonym_tvb_right

    def to_json(self, is_right_hemisphere=False, activation_patterns=None):

        children = []
        for child in self.children:
            children.append(child.to_json(is_right_hemisphere, activation_patterns))
        title = "URI: " + self.uri + "\n\nLabel: " + self.label + "\n\nDefinition: " + self.definition + \
                "\n\nSynonyms: " + self.synonym.replace("|", "\n")
        if activation_patterns is not None and str(self.id) in activation_patterns:
            connected_regions = activation_patterns[str(self.id)]
            title += "\n\nTVB " + str(len(connected_regions)) + " connected regions: " + str(connected_regions)

        if self.synonym_tvb_right >= 0 and self.synonym_tvb_left >= 0:
            # When TVB regions display differently
            synonym_id = self.synonym_tvb_right if is_right_hemisphere else self.synonym_tvb_left
            short_tvb_name = self.uri.split('#')[1]
            title = str(synonym_id) + " - " + short_tvb_name + "\n\n" + title
            return dict(data=dict(icon=ICON_TVB, title=short_tvb_name),
                        attr=dict(id=NODE_ID_TVB + str(synonym_id), title=title),
                        state="close", children=children)

        return dict(data=dict(icon=ICON_FOLDER, title=self.label.capitalize()),
                    attr=dict(id=NODE_ID_BRCO + str(self.id), title=title),
                    state="close", children=children)


class ConnectivityAnnotations(HasTraits):
    """
    Ontology annotations for a Connectivity.
    """

    connectivity = Attr(field_type=Connectivity)

    """
    Holds a flatten form for the annotations for a full connectivity.
    Each region in the connectivity can have None, or a tree of AnnotationTerms
    To be stored in a compound DS in H5.
    """
    region_annotations = NArray(
        default=numpy.array([], dtype=ANNOTATION_DTYPE),
        label="Region Annotations",
        doc="""Flat tree of annotations for every connectivity region.""")

    def set_annotations(self, annotation_terms):
        annotations = [ann.to_tuple() for ann in annotation_terms]
        annotations = numpy.array(annotations, dtype=ANNOTATION_DTYPE)
        self.region_annotations = annotations

    def summary_info(self):
        """
        Gather interesting summary information from an instance of this dataType.
        """
        summary = {"Connectivity": self.connectivity.display_name}
        summary.update(narray_summary_info(self.region_annotations, ar_name='region_annotations'))
        return summary

    def get_activation_patterns(self):
        """
        Group Annotation terms by URI.
        :return: map {brco_id: list of TVB regions IDs in which the same term is being subclass}
        """

        map_by_uri = {}
        for ann in self.region_annotations:
            ann_uri = ann[8]
            left, right = str(ann[2]), str(ann[3])
            if ann_uri not in map_by_uri:
                map_by_uri[ann_uri] = [left, right]
            else:
                if left not in map_by_uri[ann_uri]:
                    map_by_uri[ann_uri].append(left)
                if right not in map_by_uri[ann_uri]:
                    map_by_uri[ann_uri].append(right)

        map_by_brco_id = {}
        for ann in self.region_annotations:
            ann_uri = ann[8]
            ann_id = ann[0]
            map_by_brco_id[str(ann_id)] = map_by_uri[ann_uri]

        return map_by_brco_id

    def get_activation_pattern_labels(self):
        """
        :return: map {brco_id: list of TVB regions LABELS in which the same BRCO term is being subclass}
        """
        map_with_ids = self.get_activation_patterns()
        map_with_labels = dict()

        for ann_id, activated_ids in list(map_with_ids.items()):
            map_with_labels[ann_id] = []
            for string_idx in activated_ids:
                int_idx = int(string_idx)
                conn_label = self.connectivity.region_labels[int_idx]
                map_with_labels[ann_id].append(conn_label)

        return map_with_labels

    def tree_json(self):
        """
        :return: JSON to be rendered in a Tree of entities
        """
        annotations_map = dict()
        regions_map = dict()
        for i in range(self.connectivity.number_of_regions):
            regions_map[i] = []

        for ann in self.region_annotations:
            ann_obj = AnnotationTerm(ann[0], ann[1], ann[2], ann[3], ann[4], ann[5],
                                     ann[6], ann[7], ann[8], ann[9], ann[10])
            annotations_map[ann_obj.id] = ann_obj
            if ann_obj.parent_id < 0:
                # Root directly under a TVB region node
                regions_map[ann_obj.parent_left].append(ann_obj)
                regions_map[ann_obj.parent_right].append(ann_obj)
            elif ann_obj.parent_id in annotations_map:
                annotations_map[ann_obj.parent_id].add_child(ann_obj)
            else:
                self.logger.warn("Order of processing invalid parent %s child %s" % (ann_obj.parent_id, ann_obj.id))

        left_nodes, right_nodes = [], []
        activation_patterns = self.get_activation_pattern_labels()
        for region_idx, annotations_list in list(regions_map.items()):
            if_right_hemisphere = self.connectivity.is_right_hemisphere(region_idx)
            childred_json = []
            for ann_term in annotations_list:
                childred_json.append(ann_term.to_json(if_right_hemisphere, activation_patterns))
            # This node is built for every TVB region
            child_json = dict(data=dict(icon=ICON_TVB,
                                        title=self.connectivity.region_labels[region_idx]),
                              attr=dict(id=NODE_ID_TVB_ROOT + str(region_idx),
                                        title=str(region_idx) + " - " + self.connectivity.region_labels[region_idx]),
                              state="close", children=childred_json)
            if if_right_hemisphere:
                right_nodes.append(child_json)
            else:
                left_nodes.append(child_json)

        # Group everything under a single root
        left_root = dict(data=dict(title="Left Hemisphere", icon=ICON_FOLDER),
                         state="open", children=left_nodes)
        right_root = dict(data=dict(title="Right Hemisphere", icon=ICON_FOLDER),
                          state="open", children=right_nodes)
        root_root = dict(data=dict(title=self.display_name, icon=ICON_FOLDER),
                         state="open", children=[left_root, right_root])
        return root_root

File: tvb_framework/tvb/interfaces/web/templates/jinja2/visualizers/annotations/annotations.js
/**
 * TheVirtualBrain-Framework Package. This package holds all Data Management, and
 * Web-UI helpful to run brain-simulations. To use it, you also need to download
 * TheVirtualBrain-Scientific Package (for simulators). See content of the
 * documentation-folder for more details. See also http://www.thevirtualbrain.org
 *
 * (c) 2012-2023, Baycrest Centre for Geriatric Care ("Baycrest") and others
 *
 * This program is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 * PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License along with this
 * program.  If not, see <http://www.gnu.org/licenses/>.
 *
 **/

/**
 * Depends on the following GLOBALS: gl, BASE_PICK, ColSch, LEG, HLPR_readJSONfromFile
 *
 * @param treeDataUrl - URL from where the Annotations tree will be loaded
 * @param triangleToRegionUrl - URL for reading a vector of triangle to connectivity region mapping
 * @param activationPatternsUrl - URL for retrieving the Map of the activation patterns
 * @param minValue - Minimum value for the colors (used for legend)
 * @param maxValue - Maximum value for the colors (used for legend)
 * @constructor
 */
function ANN_Displayer(treeDataUrl, triangleToRegionUrl, activationPatternsUrl, minValue, maxValue) {

    this.treeElem = $("#treeStructure");

    this.triaglesMappings = [];
    this.regionToTriangleMapping = {};
    this.selectFrom3DMode = false;

    this.activationPatternMap = {};
    this.regionMappingColors = [];

    this.prefixNodeIdTVB = "node_tvb_";
    this.prefixNodeIdTVBRoot = "node_tvb_root_";
    this.prefixNodeIdBRCO = "node_brco_";

    this._init = function (treeDataUrl, triangleToRegionUrl, activationPatternsUrl, minValue, maxValue) {

        this._populateAnnotationsTree(treeDataUrl);

        ColSch_initColorSchemeGUI(minValue, maxValue);
        LEG_initMinMax(minValue, maxValue);
        LEG_generateLegendBuffers();
        BASE_PICK_initLegendInfo(maxValue, minValue);

        this.triaglesMappings = HLPR_readJSONfromFile(triangleToRegionUrl);
        this._prepareRegionToTriangleMapping();

        this.activationPatternMap = HLPR_readJSONfromFile(activationPatternsUrl);
    };

    this._populateAnnotationsTree = function (treeDataUrl) {

        this.treeElem.jstree({
            "plugins": ["themes", "json_data", "ui", "crrm"],
            "themes": {
                "theme": "default",
                "dots": true,
                "icons": true,
                "url": deploy_context + "/static/jquery/jstree-theme/style.css"
            },
            "json_data": {
                "ajax": {
                    url: treeDataUrl,
                    success: function (d) {
                        return eval(d);
                    }
                }
            }
        });

        var SELF = this;
        this.treeElem.bind("select_node.jstree", function (event, data) {

            BASE_PICK_clearFocalPoints();
            SELF._redrawColors([]);
            if (SELF.selectFrom3DMode) {
                return;
            }

            BASE_PICK_moveNavigatorToCenter();
            var selectedNode = data.rslt.obj.attr("id");
            if (selectedNode && selectedNode.indexOf(SELF.prefixNodeIdTVB) == 0) {
                selectedNode = selectedNode.replace(SELF.prefixNodeIdTVBRoot, '').replace(SELF.prefixNodeIdTVB, '');
                selectedNode = parseInt(selectedNode);
                displayMessage("Selected Region " + selectedNode, 'infoMessage');

                //TRIANGLE_pickedIndex = parseInt(SELF.regionToTriangleMapping[selectedNode]);
                //BASE_PICK_moveBrainNavigator(true);

                SELF._redrawColors([selectedNode]);

            } else if (selectedNode && selectedNode.indexOf(SELF.prefixNodeIdBRCO) == 0) {
                selectedNode = selectedNode.replace(SELF.prefixNodeIdBRCO, '');
                selectedNode = parseInt(selectedNode);

                var matchingTvbRegions = SELF.activationPatternMap[selectedNode];
                //for (var i = 0; i < matchingTvbRegions.length; i++) {
                //    var triangleIdx = parseInt(SELF.regionToTriangleMapping[matchingTvbRegions[i]]);
                //    TRIANGLE_pickedIndex = triangleIdx;
                //    BASE_PICK_moveBrainNavigator(false);
                //    BASE_PICK_addFocalPoint(triangleIdx);
                //}

                displayMessage("BRCO node connected with " + matchingTvbRegions.length +
                    " TVB regions: \n[" + matchingTvbRegions + "]", 'infoMessage');
                SELF._redrawColors(matchingTvbRegions);
            }
        })
    };

    this.selectTreeNode = function () {

        if (TRIANGLE_pickedIndex >= 0) {
            if (TRIANGLE_pickedIndex >= this.triaglesMappings.length) {
                displayMessage("Picked triangle " + TRIANGLE_pickedIndex + " outside of our region mapping", 'warningMessage');
                return;
            }
            var pickedRegion = this.triaglesMappings[TRIANGLE_pickedIndex];
            displayMessage("Picked triangle " + TRIANGLE_pickedIndex + " in region " + pickedRegion, 'infoMessage');

            this.selectFrom3DMode = true;
            this.treeElem.jstree("deselect_all");
            var toBeSelectedId = "#" + this.prefixNodeIdTVBRoot + pickedRegion;
            this.treeElem.jstree("select_node", toBeSelectedId);
            this.treeElem.jstree("open_node", toBeSelectedId);
            this.selectFrom3DMode = false;
            // disable data retrieval until next triangle is picked
            TRIANGLE_pickedIndex = GL_NOTFOUND;
        }
    };

    this.openAll = function () {
        this.treeElem.jstree('open_all');
    };

    this.closeAll = function () {
        this.treeElem.jstree('close_all');
    };

    this.setBrainColors = function (colorsUrlList) {

        colorsUrlList = $.parseJSON(colorsUrlList);
        this.regionMappingColors = [];
        for (var i = 0; i < colorsUrlList.length; i++) {
            var oneData = HLPR_readJSONfromFile(colorsUrlList[i]);
            this.regionMappingColors.push(oneData);
        }
        this._redrawColors([]);
    };

    this._redrawColors = function (activeRegions) {

        if (!activeRegions || activeRegions.length == 0) {
            BASE_PICK_updateBrainColors(this.regionMappingColors);
            return;
        }

        var currentColors = [];
        for (var i = 0; i < this.regionMappingColors.length; i++) {
            var chunkColors = this.regionMappingColors[i].slice(0);
            for (var j = 0; j < chunkColors.length; j++) {
                var isActive = false;
                for (var k = 0; k < activeRegions.length; k++) {
                    if (activeRegions[k] == chunkColors[j]) {
                        isActive = true;
                    }
                }
                if (!isActive) {
                    chunkColors[j] = -1;
                }
            }
            currentColors.push(chunkColors);
        }
        BASE_PICK_updateBrainColors(currentColors);
    };


    this._prepareRegionToTriangleMapping = function () {

        for (var i = 0; i < this.triaglesMappings.length; i++) {
            var region = this.triaglesMappings[i];
            if (this.regionToTriangleMapping[region]) {
                continue;
            }
            this.regionToTriangleMapping[region] = i;
        }
    };

    this._init(treeDataUrl, triangleToRegionUrl, activationPatternsUrl, minValue, maxValue);
}